modules/apps/callbacks/ibc_middleware.go

```diff
@@ -0,0 +1,309 @@
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback fails (within the gas limit), state changes are reverted.
+func (im IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	// we first call the underlying app to handle the acknowledgement
+	err := im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	if err != nil {
+		return err
+	}
+
+	packetSenderAddress := im.GetPacketSender(packet)
+	callbackDataGetter := func() (types.CallbackData, bool, error) {
+		return types.GetSourceCallbackData(im.app, packet.Data, ctx.GasMeter().GasRemaining(), im.maxCallbackGas)
+	}
+	callbackExecutor := func(cachedCtx sdk.Context, callbackAddress string) error {
+		return im.contractKeeper.IBCOnAcknowledgementPacketCallback(cachedCtx, packet, acknowledgement, relayer, callbackAddress, packetSenderAddress)
+	}
+
+	return im.processCallback(ctx, packet, types.CallbackTypeAcknowledgement, callbackDataGetter, callbackExecutor)
+}
+
+// OnTimeoutPacket implements timeout source callbacks for the ibc-callbacks middleware.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback fails (within the gas limit), state changes are reverted.
+func (im IBCMiddleware) OnTimeoutPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) error {
+	err := im.app.OnTimeoutPacket(ctx, packet, relayer)
+	if err != nil {
+		return err
+	}
+
+	packetSenderAddress := im.GetPacketSender(packet)
+	callbackDataGetter := func() (types.CallbackData, bool, error) {
+		return types.GetSourceCallbackData(im.app, packet.Data, ctx.GasMeter().GasRemaining(), im.maxCallbackGas)
+	}
+	callbackExecutor := func(cachedCtx sdk.Context, callbackAddress string) error {
+		return im.contractKeeper.IBCOnTimeoutPacketCallback(cachedCtx, packet, relayer, callbackAddress, packetSenderAddress)
+	}
+
+	return im.processCallback(ctx, packet, types.CallbackTypeTimeoutPacket, callbackDataGetter, callbackExecutor)
+}
+
+// OnRecvPacket implements the WriteAcknowledgement destination callbacks for the ibc-callbacks middleware during
+// synchronous packet acknowledgement.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback fails (within the gas limit), state changes are reverted via a panic.
+func (im IBCMiddleware) OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) ibcexported.Acknowledgement {
+	ack := im.app.OnRecvPacket(ctx, packet, relayer)
+	// if ack is nil, then the callback is handled in WriteAcknowledgement
+	if ack == nil {
+		return nil
+	}
+
+	packetReceiverAddress := im.GetPacketReceiver(packet)
+	callbackDataGetter := func() (types.CallbackData, bool, error) {
+		return types.GetDestCallbackData(im.app, packet.GetData(), ctx.GasMeter().GasRemaining(), im.maxCallbackGas)
```

Comment:
Super nit: you call `GetData` here, but access `Data` in the other functions.

```suggestion
		return types.GetDestCallbackData(im.app, packet.Data, ctx.GasMeter().GasRemaining(), im.maxCallbackGas)
```

Reply:
changed all to `GetData()`